funnel/transports/sms.py

```diff
@@ -9,65 +17,131 @@
+        elif phone_number.startswith('+'):
+            return self._route_via_twilio(phone_number, message)
+        raise TransportRecipientError(
+            "No service provider available for this recipient"
         )
-        if r.status_code in (200, 201):
-            # All good
-            jsonresponse = r.json()
-            return jsonresponse.get('sid')
-        raise TransportTransactionError("Twilio API error", r.status_code, r.text)
-    except requests.ConnectionError:
-        raise TransportConnectionError("Twilio not reachable")
+
+    @staticmethod
+    def _route_via_exotel(phone_number: str, message: str) -> str:
+        """
+        Route the SMS via Twilio. This is done only for messages that can't be delivered by exotel.
+        :param phone_number: Phone Number
+        :param message: Message to deliver to Phone number.
+        :return SMS Message ID
+        """
+        sid = app.config['SMS_EXOTEL_SID']
+        token = app.config['SMS_EXOTEL_TOKEN']
+        try:
+            r = requests.post(
+                'https://twilix.exotel.in/v1/Accounts/{sid}/Sms/send.json'.format(
+                    sid=sid
+                ),
+                auth=(sid, token),
+                data={
+                    'From': app.config['SMS_EXOTEL_FROM'],
+                    'To': phone_number,
+                    'Body': message,
+                },
+            )
+            if r.status_code in (200, 201):
+                # All good
+                jsonresponse = r.json()
+                if isinstance(jsonresponse, (list, tuple)) and jsonresponse:
+                    transactionid = jsonresponse[0].get('SMSMessage', {}).get('Sid')
+                else:
+                    transactionid = jsonresponse.get('SMSMessage', {}).get('Sid')
+                return transactionid
+            raise TransportTransactionError("Exotel API error", r.status_code, r.text)
+        except requests.ConnectionError:
+            raise TransportConnectionError("Exotel not reachable")
+
+    def _route_via_twilio(self, phone: str, message: str) -> str:
+        """
+        Route the SMS via Twilio
+        :param phone: Phone Number
+        :param message: Message to deliver to Phone number.
+        :return SMS Message ID
+        """
+        # Get SID, Token and From (these are required to make any calls)
+        account = app.config['SMS_TWILIO_SID']
+        token = app.config['SMS_TWILIO_TOKEN']
+        sender = app.config['SMS_TWILIO_FROM']
+
+        # Send (This uses the routing API to deliver SMS via a Low Latency Location).
+        # See https://www.twilio.com/docs/global-infrastructure/edge-locations
+        client = Client(account, token)
+
+        # We need a callback url, if it was not generated already.
+        if (not self.twilio_callback) and self.callback:
+            with app.app_context():
+                self.twilio_callback = url_for('process_twilio_event', _external=True)
```

Comment:
Don't use `with app.app_context()` when you already have a context (such as when this is called from a view). Database sessions are linked to app context, so creating one will invalidate the existing database session and all ORM objects loaded from it, plus the ongoing DB transaction. The calling function will be unable to use the database after this.

Always use `with app.app_context()` as the _outermost_ construct. We learnt this the hard way a few months ago, resulting in the refactoring in ff99ff22c8a7e7cb4b524a4f45d3482f7ccd7efa.

Reply:
You can use `app_url_for` here, but it's not needed. The problem is in the tests, not here.